app/validators/qc_receptions_factory_validator.rb

```diff
@@ -0,0 +1,42 @@
+# frozen_string_literal: true
+
+# Failed validations return unprocessable_entity
+# These are being validated before any entity is created
+class QcReceptionsFactoryValidator < ActiveModel::Validator
+  def validate(record)
+    validations = %i[validate_qc_results_list validate_assay_type]
+
+    validations.each do |validation|
+      next if record.errors.present?
+
+      send(validation, record)
+    end
+  end
+
+  def validate_qc_results_list(record)
+    # Check if array is not empty
+    if record.qc_results_list.empty?
+      record.errors.add :qc_results_list, :empty_array
+      return
+    end
+    # Return if any object inside the qc_results_list array has values
+    return nil if record.qc_results_list.any? { |value| !value.empty? }
+
+    # Add error if the objects inside the qc_results_list array is empty
+    record.errors.add :qc_results_list, :empty_array
+    nil
+  end
+
+  def validate_assay_type(record)
+    # Iterates through the qc_results_list array objects
+    # checks if any qc attribute matches the assay type key
+    # add error if not
+    assay_types = QcAssayType.where(used_by: options[:used_by]).pluck(:id, :key)
+    assay_types_hash = assay_types.to_h.invert
+    record.qc_results_list.each do |qc_hash|
+      return nil if qc_hash.any? { |qc, _value| assay_types_hash.keys.include? qc }
+    end
+    record.errors.add :qc_results_list, :invalid
+    nil
+  end
```

Comment:
Again, this could be refactored as nil is being returned in two places. Which makes you think if you inverse the check and return the error. Then otherwise, return nil

Reply:
Tried an inverse of the check, did not quite work well. 
I just want to check if there is at least one right matching qc_assay_type in  any of the hashes inside the array.